### PR TITLE
[CPP Graph] Falcon 40B

### PR DESCRIPTION
**Pull request by zhentaoyu**
## Type of Change

feature or bug fix or documentation or others
API changed or not: no
support [Falcon-40b](https://huggingface.co/tiiuae/falcon-40b)

## Description

detail description 
JIRA ticket: [822](https://jira.devtools.intel.com/browse/NLPTOOLKIU-822)
TODO
- [x] `n_head_kv`
- [x] arch changes
- [x] kernel parts (group broadcast gemm? like `128 x 1 x 64` x `8 x 64 x 32` or repeat (memcpy) 
- [x] extension tests 

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR 

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

**Comment by zhentaoyu** [comment]
logits output comparisons (first token):
`prompt: What is the meaning of life?`
`PyTorch` (use `torch.bfloat16` for acceleration):
```bash
inputs {'input_ids': tensor([[1562,  304,  248, 4113,  275, 1063,   42]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1]])}
(Pdb) out['logits'].shape
torch.Size([1, 7, 65024])
(Pdb) out['logits'][0][-1][:32]
tensor([ -7.6562, -14.6250, -17.2500, -16.2500,  -1.8594,  -8.6250,  -9.8125,
         -4.7500, -22.2500, -22.1250, -22.1250,  -3.9062,  -5.9375,  -5.9688,
         -4.9375,  -5.8438,  -6.2188,  -6.2500,  -6.6562,  -5.2188,  -8.4375,
         -1.8047,  -6.7812,  -5.3125,  -2.3906,  -5.5938,  -0.4941,  -4.2812,
         -2.6406,  -3.4531,  -3.7344,  -4.0312], dtype=torch.bfloat16,
       grad_fn=<SliceBackward0>)
```

`NE Graph`
- f32:
  ```bash
  numactl -l -C 56-111 ./build/bin/chat_falcon -m fp32.bin -t 56 -c 2048 -b 1024 -n 32 -p "What is the meaning of life?" --seed 
  16 
  What is the meaning of life? input_ids: 1562, 304, 248, 4113, 275, 1063, 42,
  logits: -7.63562, -14.62637, -17.17973, -16.28150, -1.80734, -8.57081, -9.76584, -4.69932, -22.37857, -22.37496, -22.36914, 
  -3.78765, -5.94660, -5.96269, -4.94051, -5.84046, -6.22719, -6.24325, -6.64994, -5.23302, -8.42179, -1.78704, -6.80018, 
  -5.31788, -2.38076, -5.61912, -0.49721, -4.26610, -2.63580, -3.43246, -3.71723, -4.03147,
  ```

- q4_0:
  ```bash
  numactl -l -C 56-111 ./build/bin/chat_falcon -m fp32.bin -t 56 -c 2048 -b 1024 -n 32 -p "What is the meaning of life?" --seed 
  16 
  What is the meaning of life? input_ids: 1562, 304, 248, 4113, 275, 1063, 42,
  logits: -7.56420, -13.86204, -16.06506, -15.73748, -1.35945, -7.86627, -9.36824, -4.14623, -22.71499, -22.75335, -22.75961, 
  -3.15193, -5.55903, -5.72819, -4.93742, -5.70148, -5.18694, -6.00259, -6.42102, -5.34242, -8.70160, -1.21200, -6.78930, 
  -4.98841, -2.19866, -5.23280, -0.57787, -4.52719, -2.26344, -3.56918, -3.31708, -3.99321,
  ```

**Comment by zhentaoyu** [comment]
`numactl -l -C 56-111 ./build/bin/chat_falcon -m q4_0.bin -t 56 -c 2048 -b 1024 -n 1024 -p "What is the meaning of life?" --seed 16`
```bash
What is the meaning of life?
The meaning of life is whatever you make it to be. The meaning of life is different for each person. There is no one answer to this question. The meaning of life is different for each person, so everyone must find their own meaning. There is no right or wrong answer.
There are many different ways to look at the meaning of life. Some people believe that it means working hard to achieve goals and make a difference in the world. Others believe that the meaning of life is simply living each day to its fullest. Still others believe that the meaning of life
 is finding happiness and enjoying each moment.
The meaning of life is whatever you make it to be. You can make your life meaningful by living each day to its fullest, working hard to achieve your goals, and finding happiness in the journey. No matter what the meaning of life is for you, make sure that you live a life that is worth liv
ing.
The meaning of life is different for each person. For some, it might be the search for truth. For others, it might be to help others. And for still others, it might be to enjoy life and experience all that it has to offer. No matter what the meaning of life is for you, take some time to t
hink about it and figure out what makes your life meaningful.
The meaning of life is whatever you want it to be. You can make your life meaningful by living each day to its fullest, working hard to achieve your goals, and finding happiness in the journey.<|endoftext|>If you’re struggling to find the meaning of life, you’re not alone.
Many people feel lost and confused when they try to figure out what their purpose is in life. It can be difficult to find the meaning of life when we don’t know what our purpose is. However, there are some things that can help us find our way.
One way to find the meaning of life is to focus on what you love. What are the things that bring you joy? What are the things that make you feel alive? What are the things that you’re passionate about? When you focus on what you love, you’ll be able to figure out what your purpose is in l
ife.
Another way to find the meaning of life is to think about what you want to do with your life. What is your dream job? What do you want to achieve in life? What do you want to leave behind? When you think about what you want to do with your life, you’ll be able to figure out what your purp
ose is in life.
The Meaning of Life is a question that has been asked by philosophers for centuries. Some have said that the meaning of life is to enjoy it while we have it, while others have said that the meaning of life is to find our purpose and live it out.<|endoftext|>The Meaning of Life is a questi
on that has been asked by philosophers for centuries. Some have said that the meaning of life is to enjoy it while we have it, while others have said that the meaning of life is to find our purpose and live it out.
There is no definitive answer to the question, “What is the meaning of life?” There are many different theories, but none can be proven absolutely true. However, one theory that has been gaining popularity in recent years is the idea that the meaning of life is to find your purpose and li
ve it out.
This theory suggests that each person has a unique purpose in life that they must fulfill in order to find happiness. The purpose could be anything from being a good parent to finding love. Whatever it is, the key is to identify your purpose and work towards achieving it.
If you can find your purpose and live it out, then you will find meaning in your life. This may sound like a simple concept, but it can be difficult to put into practice. It takes time and effort to identify your purpose and then work towards achieving it. However, if you are able to do t
his, then you will find meaning in your life.
The meaning of life is different for everyone. Some people believe that the meaning of life is to find happiness, while others believe that the meaning of life is to find God. Whatever the meaning of life is, it is important to find your own meaning and live your life accordingly.
The search for the meaning of life has been a central theme in human thought for thousands of years. While there is no one definitive answer, many people believe that the meaning of life is to be happy. Happiness can come in many forms, such as achieving goals, connecting with others, or
simply enjoying the present moment.
Happiness is often associated with pleasure, but it does not have to be. Happiness can also come from finding meaning in one’s life, whether through a career, hobbies, or relationships. Whatever form happiness takes, the ultimate goal is to find purpose and fulfillment in life.
The meaning of life is different for everyone. Some people believe that the meaning of life is to find happiness, while
```

**Comment by zhentaoyu** [comment]
perf for reference (1ins / 56 cores, M6)
| out_new_tokens |  fist_token latency (ms) | second_token_latency (ms) |
| ------------------| --------------------------| ------------------------------ |
| 32                      |          674.33                   |            169.73                       |
|  256                   |           663.3                    |            170.11                       |
| 1024                  |            643.40                 |           162.59                        |

**Comment by zhentaoyu** [comment]
https://inteltf-jenk.sh.intel.com/view/nlp-toolkit-validation/job/ITREX-cpp-graph-extension/64/

**Review by a32543254 — APPROVED** [review]
LGTM

**Review by zhenwei-intel — APPROVED** [review]
LGTM
